go/stellar/time_cache.go

```diff
@@ -0,0 +1,124 @@
+package stellar
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/keybase/client/go/libkb"
+)
+
+// Threadsafe cache with expiration and singleflighting.
+type TimeCache struct {
+	name     string // name for logging
+	maxAge   time.Duration
+	initLock sync.Mutex
+	lockTab  libkb.LockTable
+	cache    *lru.Cache
+}
+
+type timeCacheEntry struct {
+	val  interface{}
+	time time.Time
+}
+
+func NewTimeCache(name string, size int, maxAge time.Duration) *TimeCache {
+	if size <= 0 {
+		size = 10
+	}
+	cache, err := lru.New(size)
+	if err != nil {
+		panic(err)
+	}
+	return &TimeCache{
+		name:   name,
+		maxAge: maxAge,
+		cache:  cache,
+	}
+}
+
+func (c *TimeCache) init() {
+	c.initLock.Lock()
+	defer c.initLock.Unlock()
+	if c.cache != nil {
+		return
+	}
+	cache, err := lru.New(50)
+	if err != nil {
+		panic(err)
+	}
+	c.cache = cache
+}
+
+type cacheFillFunc = func() (interface{}, error)
+
+func (c *TimeCache) Get(mctx libkb.MetaContext, key string, into interface{}) (ok bool) {
+	return c.getHelper(mctx, key, into, nil) == nil
+}
+
+// GetWithFill is prefereable to Get because it holds a locktab lock during the fill.
+// Which prevents concurrent accesses from doing the extra work of running fill at the same time.
+func (c *TimeCache) GetWithFill(mctx libkb.MetaContext, key string, into interface{}, fill cacheFillFunc) error {
+	return c.getHelper(mctx, key, into, fill)
+}
+
+func (c *TimeCache) getHelper(mctx libkb.MetaContext, key string, into interface{}, fill cacheFillFunc) error {
+	lock, err := c.lockTab.AcquireOnNameWithContextAndTimeout(mctx.Ctx(), mctx.G(), key, time.Minute)
+	if err != nil {
+		return fmt.Errorf("could not acquire cache lock for key %v", key)
+	}
+	defer lock.Release(mctx.Ctx())
+	storeResult := func(val interface{}) (ok bool) {
```

Comment:
not a huge deal, but preference for this to be a top-level function that takes two parameters. not that i think it'll be re-used, just a little easier to follow this code path and an opportunity for more obvious naming.

Reply:
So `Get(mctx, key, into, fill) error` where `fill` is optional? The thing I think is weird is this. If you _do_ pass `fill` then the `error` returned should be treated as a failure and likely returned from the caller. But if `fill` is not passed then `error` is just a cache miss.

Reply:
just storeResult. storeResult doesn't call `fill`. it puts the thing in the thing if it can. just that.

Reply:
oh that, yeah sounds good